File: evernote_client/client/queue.py
"""Persistent write-operation queue backed by SQLite."""

import logging
from collections.abc import Callable
from pathlib import Path
from typing import Any

import persistqueue

logger = logging.getLogger(__name__)


class OperationQueue:
    def __init__(self, path: Path) -> None:
        path.mkdir(parents=True, exist_ok=True, mode=0o700)
        self._q: persistqueue.SQLiteQueue = persistqueue.SQLiteQueue(
            str(path), auto_commit=True
        )

    MAX_RETRIES = 5

    def put(self, operation: str, **params: Any) -> None:
        self._q.put({"operation": operation, "params": params, "retries": 0})

    def process_all(self, dispatcher: dict[str, Callable[..., Any]]) -> list[Any]:
        """Process all queued operations.

        Failed operations are logged and re-enqueued so they are not lost.
        Returns results of successful operations.
        """
        results = []
        failed: list[dict[str, Any]] = []
        while not self._q.empty():
            item = self._q.get(block=False)
            operation = item["operation"]
            fn = dispatcher.get(operation)
            if fn is None:
                logger.error("Unknown queued operation %r — dropping", operation)
                self._q.task_done()
                continue
            try:
                results.append(fn(**item["params"]))
                self._q.task_done()
            except Exception:
                retries = item.get("retries", 0) + 1
                if retries >= self.MAX_RETRIES:
                    logger.error(
                        "Queued operation %r failed %d times — dropping",
                        operation,
                        retries,
                    )
                else:
                    logger.warning(
                        "Queued operation %r failed — will re-enqueue (attempt %d/%d)",
                        operation,
                        retries,
                        self.MAX_RETRIES,
                    )
                    logger.debug(
                        "Queued operation %r failure details",
                        operation,
                        exc_info=True,
                    )
                    failed.append({**item, "retries": retries})
                self._q.task_done()
        for item in failed:
            self._q.put(item)
        return results

    def size(self) -> int:
        return self._q.size

    def is_empty(self) -> bool:
        return self._q.empty()

File: src/client/queue.ts
import { mkdirSync, existsSync, readFileSync, writeFileSync } from 'node:fs';
import { chmod } from 'node:fs/promises';
import { join } from 'node:path';
import { logger } from '../logger.js';

interface QueueItem {
  operation: string;
  params: Record<string, unknown>;
  retries: number;
}

type Dispatcher = Record<
  string,
  (params: Record<string, unknown>) => Promise<unknown>
>;

const MAX_RETRIES = 5;

/**
 * Persistent write-operation queue backed by a JSON file.
 * Simple file-based queue for durability without requiring SQLite.
 */
export class OperationQueue {
  private filePath: string;

  constructor(queuePath: string) {
    mkdirSync(queuePath, { recursive: true });
    chmod(queuePath, 0o700).catch((err) =>
      logger.warn(`Could not set queue dir permissions: ${err}`)
    );
    this.filePath = join(queuePath, 'queue.json');
  }

  private load(): QueueItem[] {
    if (!existsSync(this.filePath)) return [];
    try {
      return JSON.parse(readFileSync(this.filePath, 'utf-8'));
    } catch {
      return [];
    }
  }

  private save(items: QueueItem[]): void {
    writeFileSync(this.filePath, JSON.stringify(items, null, 2));
  }

  put(operation: string, params: Record<string, unknown>): void {
    const items = this.load();
    items.push({ operation, params, retries: 0 });
    this.save(items);
  }

  async processAll(dispatcher: Dispatcher): Promise<unknown[]> {
    const items = this.load();
    if (items.length === 0) return [];

    const results: unknown[] = [];
    const failed: QueueItem[] = [];

    for (const item of items) {
      const fn = dispatcher[item.operation];
      if (!fn) {
        logger.error(`Unknown queued operation '${item.operation}' — dropping`);
        continue;
      }
      try {
        results.push(await fn(item.params));
      } catch (err) {
        logger.warn({ err }, `Queued operation '${item.operation}' error`);
        const retries = (item.retries ?? 0) + 1;
        if (retries >= MAX_RETRIES) {
          logger.error(
            `Queued operation '${item.operation}' failed ${retries} times — dropping`
          );
        } else {
          logger.warn(
            `Queued operation '${item.operation}' failed — will re-enqueue (attempt ${retries}/${MAX_RETRIES})`
          );
          failed.push({ ...item, retries });
        }
      }
    }

    this.save(failed);
    return results;
  }

  size(): number {
    return this.load().length;
  }

  isEmpty(): boolean {
    return this.size() === 0;
  }
}
